File: worldgenconverter.cpp
// molang_to_density.cpp
// Compile with: g++ -std=c++17 molang_to_density.cpp -o molang_to_density
// this doesnt work because we need  nlohmann/json available as "json.hpp"

#include <iostream>
#include <string>
#include <vector>
#include <stdexcept>
#include <sstream>
#include <memory>
#include <cctype>

// Include nlohmann::json library (download from https://github.com/nlohmann/json)
#include "json.hpp"
using json = nlohmann::json;

using namespace std;

//////////////////////////////////////////////
// TOKENIZER
//////////////////////////////////////////////

struct Token {
    string type;
    string value;
};

vector<Token> tokenize(const string &s) {
    vector<Token> tokens;
    size_t i = 0;
    while (i < s.size()) {
        char c = s[i];
        if (isspace(c)) {
            i++;
            continue;
        }
        if (isdigit(c) || (c == '.' && i+1 < s.size() && isdigit(s[i+1]))) {
            string numStr;
            numStr.push_back(c);
            i++;
            while (i < s.size() && (isdigit(s[i]) || s[i]=='.')) {
                numStr.push_back(s[i]);
                i++;
            }
            tokens.push_back({ "NUMBER", numStr });
            continue;
        }
        if (isalpha(c) || c=='_' || c=='.') {
            string idStr;
            idStr.push_back(c);
            i++;
            while (i < s.size() && (isalnum(s[i]) || s[i]=='_' || s[i]=='.')) {
                idStr.push_back(s[i]);
                i++;
            }
            tokens.push_back({ "IDENTIFIER", idStr });
            continue;
        }
        // Multi-character operators for <, >, !, =
        if (c=='<' || c=='>' || c=='!' || c=='=') {
            string op;
            op.push_back(c);
            if (i+1 < s.size() && s[i+1]=='=') {
                op.push_back('=');
                i += 2;
            } else {
                i++;
            }
            tokens.push_back({ "OPERATOR", op });
            continue;
        }
        if (c=='+' || c=='-' || c=='*' || c=='/' || c=='%') {
            tokens.push_back({ "OPERATOR", string(1,c) });
            i++;
            continue;
        }
        if (c=='(' || c==')') {
            tokens.push_back({ "PAREN", string(1,c) });
            i++;
            continue;
        }
        if (c==',') {
            tokens.push_back({ "COMMA", "," });
            i++;
            continue;
        }
        if (c=='?') {
            tokens.push_back({ "QUESTION", "?" });
            i++;
            continue;
        }
        if (c==':') {
            tokens.push_back({ "COLON", ":" });
            i++;
            continue;
        }
        if (c==';') {
            tokens.push_back({ "SEMICOLON", ";" });
            i++;
            continue;
        }
        if (c=='{') {
            tokens.push_back({ "LCURLY", "{" });
            i++;
            continue;
        }
        if (c=='}') {
            tokens.push_back({ "RCURLY", "}" });
            i++;
            continue;
        }
        // Skip any unrecognized character
        i++;
    }
    tokens.push_back({ "EOF", "" });
    return tokens;
}

//////////////////////////////////////////////
// AST NODE CLASSES
//////////////////////////////////////////////

class ASTNode {
public:
    virtual string toMolang() = 0;
    virtual json toDensityJson() = 0;
    virtual ~ASTNode() {}
};

using ASTNodePtr = unique_ptr<ASTNode>;

class NumberNode : public ASTNode {
public:
    double value;
    NumberNode(const string &valStr) {
        value = stod(valStr);
    }
    string toMolang() override {
        return to_string(value);
    }
    json toDensityJson() override {
        return json{{"type", "minecraft:number"}, {"value", value}};
    }
};

class VariableNode : public ASTNode {
public:
    string name;
    VariableNode(const string &n) : name(n) {}
    string toMolang() override {
        return name;
    }
    json toDensityJson() override {
        return json{{"type", "minecraft:variable"}, {"name", name}};
    }
};

class BinaryOpNode : public ASTNode {
public:
    string op;
    ASTNodePtr left, right;
    BinaryOpNode(string op, ASTNodePtr left, ASTNodePtr right)
        : op(op), left(move(left)), right(move(right)) {}
    string toMolang() override {
        return "(" + left->toMolang() + " " + op + " " + right->toMolang() + ")";
    }
    json toDensityJson() override {
        return json{{"type", "minecraft:binary_operation"},
                    {"operator", op},
                    {"left", left->toDensityJson()},
                    {"right", right->toDensityJson()}};
    }
};

class FunctionCallNode : public ASTNode {
public:
    string functionName;
    vector<ASTNodePtr> arguments;
    FunctionCallNode(const string &fname) : functionName(fname) {}
    void addArgument(ASTNodePtr arg) {
        arguments.push_back(move(arg));
    }
    string toMolang() override {
        string result = functionName + "(";
        for (size_t i = 0; i < arguments.size(); i++) {
            result += arguments[i]->toMolang();
            if(i != arguments.size()-1)
                result += ", ";
        }
        result += ")";
        return result;
    }
    json toDensityJson() override {
        // Map specific functions
        if(functionName=="q.noise" && arguments.size()==2) {
            return json{{"type", "minecraft:noise"},
                        {"x", arguments[0]->toDensityJson()},
                        {"y", arguments[1]->toDensityJson()}};
        } else if(functionName=="math.clamp" && arguments.size()==3) {
            return json{{"type", "minecraft:clamp"},
                        {"input", arguments[0]->toDensityJson()},
                        {"min", arguments[1]->toDensityJson()},
                        {"max", arguments[2]->toDensityJson()}};
        } else if(functionName=="math.floor" && arguments.size()==1) {
            return json{{"type", "minecraft:floor"},
                        {"input", arguments[0]->toDensityJson()}};
        } else if(functionName=="math.max" && arguments.size()==2) {
            return json{{"type", "minecraft:max"},
                        {"a", arguments[0]->toDensityJson()},
                        {"b", arguments[1]->toDensityJson()}};
        }
        // Generic function node
        json argsJson = json::array();
        for(auto &arg : arguments)
            argsJson.push_back(arg->toDensityJson());
        return json{{"type", "minecraft:function"},
                    {"name", functionName},
                    {"args", argsJson}};
    }
};

class TernaryOpNode : public ASTNode {
public:
    ASTNodePtr condition, trueExpr, falseExpr;
    TernaryOpNode(ASTNodePtr cond, ASTNodePtr tExpr, ASTNodePtr fExpr)
        : condition(move(cond)), trueExpr(move(tExpr)), falseExpr(move(fExpr)) {}
    string toMolang() override {
        return "(" + condition->toMolang() + " ? " + trueExpr->toMolang() + " : " + falseExpr->toMolang() + ")";
    }
    json toDensityJson() override {
        return json{{"type", "minecraft:ternary"},
                    {"condition", condition->toDensityJson()},
                    {"true", trueExpr->toDensityJson()},
                    {"false", falseExpr->toDensityJson()}};
    }
};

class AssignmentNode : public ASTNode {
public:
    unique_ptr<VariableNode> variable;
    ASTNodePtr expression;
    AssignmentNode(unique_ptr<VariableNode> var, ASTNodePtr expr)
        : variable(move(var)), expression(move(expr)) {}
    string toMolang() override {
        return variable->toMolang() + " = " + expression->toMolang();
    }
    json toDensityJson() override {
        return json{{"type", "minecraft:assign"},
                    {"variable", variable->toDensityJson()},
                    {"value", expression->toDensityJson()}};
    }
};

class ReturnNode : public ASTNode {
public:
    ASTNodePtr expression;
    ReturnNode(ASTNodePtr expr) : expression(move(expr)) {}
    string toMolang() override {
        return "return " + expression->toMolang();
    }
    json toDensityJson() override {
        return json{{"type", "minecraft:return"},
                    {"value", expression->toDensityJson()}};
    }
};

class BlockNode : public ASTNode {
public:
    vector<ASTNodePtr> statements;
    void addStatement(ASTNodePtr stmt) {
        statements.push_back(move(stmt));
    }
    string toMolang() override {
        string res = "{ ";
        for (size_t i = 0; i < statements.size(); i++) {
            res += statements[i]->toMolang();
            if(i < statements.size()-1)
                res += "; ";
        }
        res += " }";
        return res;
    }
    json toDensityJson() override {
        json arr = json::array();
        for(auto &stmt: statements)
            arr.push_back(stmt->toDensityJson());
        return json{{"type", "minecraft:block"}, {"statements", arr}};
    }
};

class LoopNode : public ASTNode {
public:
    ASTNodePtr iterations;
    unique_ptr<BlockNode> body;
    LoopNode(ASTNodePtr iter, unique_ptr<BlockNode> b)
        : iterations(move(iter)), body(move(b)) {}
    string toMolang() override {
        return "loop(" + iterations->toMolang() + ", " + body->toMolang() + ")";
    }
    json toDensityJson() override {
        return json{{"type", "minecraft:loop"},
                    {"iterations", iterations->toDensityJson()},
                    {"body", body->toDensityJson()}};
    }
};

//////////////////////////////////////////////
// PARSER
//////////////////////////////////////////////

class Parser {
public:
    vector<Token> tokens;
    size_t pos = 0;
    
    Parser(const vector<Token> &tokens) : tokens(tokens) { }
    
    Token current() {
        return tokens[pos];
    }
    
    void advance() {
        if(pos < tokens.size()-1)
            pos++;
    }
    
    void expect(const string &type, const string &value="") {
        Token tok = current();
        if(tok.type != type || (!value.empty() && tok.value != value))
            throw runtime_error("Expected token " + type + " " + value + " but got " + tok.type + " " + tok.value);
        advance();
    }
    
    // program = { statement ";" }+
    unique_ptr<BlockNode> parseProgram() {
        auto block = make_unique<BlockNode>();
        while(current().type != "EOF") {
            auto stmt = parseStatement();
            block->addStatement(move(stmt));
            if(current().type == "SEMICOLON")
                advance();
        }
        return block;
    }
    
    // statement: "return" expression | "loop" "(" expression "," block ")" | expression (and check for assignment)
    ASTNodePtr parseStatement() {
        if(current().type == "IDENTIFIER" && current().value == "return") {
            advance();
            return make_unique<ReturnNode>(parseExpression());
        }
        if(current().type == "IDENTIFIER" && current().value == "loop") {
            return parseLoopStatement();
        }
        // Otherwise, parse an expression. If it's an assignment, handle it.
        auto expr = parseExpression();
        if(dynamic_cast<VariableNode*>(expr.get()) != nullptr &&
           current().type == "OPERATOR" && current().value == "=") {
            // Assignment: variable "=" expression
            auto varNode = make_unique<VariableNode>(dynamic_cast<VariableNode*>(expr.release())->name);
            advance(); // consume '='
            auto rightExpr = parseExpression();
            return make_unique<AssignmentNode>(move(varNode), move(rightExpr));
        }
        return expr;
    }
    
    ASTNodePtr parseLoopStatement() {
        // assume current is "loop"
        advance(); // consume "loop"
        expect("PAREN", "(");
        auto iterations = parseExpression();
        expect("COMMA", ",");
        auto body = parseBlock();
        expect("PAREN", ")");
        return make_unique<LoopNode>(move(iterations), move(body));
    }
    
    unique_ptr<BlockNode> parseBlock() {
        expect("LCURLY", "{");
        auto block = make_unique<BlockNode>();
        while(current().type != "RCURLY") {
            auto stmt = parseStatement();
            block->addStatement(move(stmt));
            if(current().type == "SEMICOLON")
                advance();
        }
        expect("RCURLY", "}");
        return block;
    }
    
    // expression = ternary
    ASTNodePtr parseExpression() {
        return parseTernary();
    }
    
    // ternary = comparison ("?" expression ":" expression)?
    ASTNodePtr parseTernary() {
        auto cond = parseComparison();
        if(current().type == "QUESTION") {
            advance();
            auto trueExpr = parseExpression();
            expect("COLON", ":");
            auto falseExpr = parseExpression();
            return make_unique<TernaryOpNode>(move(cond), move(trueExpr), move(falseExpr));
        }
        return cond;
    }
    
    // comparison = additive { ( "<" | ">" | "<=" | ">=" | "==" | "!=" ) additive }
    ASTNodePtr parseComparison() {
        auto node = parseAdditive();
        while(current().type == "OPERATOR" && 
              (current().value=="<" || current().value==">" || current().value=="<=" ||
               current().value==">=" || current().value=="==" || current().value=="!=")) {
            string op = current().value;
            advance();
            auto right = parseAdditive();
            node = make_unique<BinaryOpNode>(op, move(node), move(right));
        }
        return node;
    }
    
    // additive = multiplicative { ("+" | "-") multiplicative }
    ASTNodePtr parseAdditive() {
        auto node = parseMultiplicative();
        while(current().type == "OPERATOR" && (current().value=="+" || current().value=="-")) {
            string op = current().value;
            advance();
            auto right = parseMultiplicative();
            node = make_unique<BinaryOpNode>(op, move(node), move(right));
        }
        return node;
    }
    
    // multiplicative = unary { ("*" | "/" | "%") unary }
    ASTNodePtr parseMultiplicative() {
        auto node = parseUnary();
        while(current().type == "OPERATOR" && (current().value=="*" || current().value=="/" || current().value=="%")) {
            string op = current().value;
            advance();
            auto right = parseUnary();
            node = make_unique<BinaryOpNode>(op, move(node), move(right));
        }
        return node;
    }
    
    // unary = ("+" | "-") unary | primary
    ASTNodePtr parseUnary() {
        if(current().type == "OPERATOR" && (current().value=="+" || current().value=="-")) {
            string op = current().value;
            advance();
            auto node = parseUnary();
            // Represent unary as (0 op expr)
            return make_unique<BinaryOpNode>(op, make_unique<NumberNode>("0"), move(node));
        }
        return parsePrimary();
    }
    
    // primary = NUMBER | IDENTIFIER (function call?) | "(" expression ")" | block
    ASTNodePtr parsePrimary() {
        Token tok = current();
        if(tok.type == "NUMBER") {
            advance();
            return make_unique<NumberNode>(tok.value);
        }
        if(tok.type == "IDENTIFIER") {
            string id = tok.value;
            advance();
            if(current().type == "PAREN" && current().value=="(")
                return parseFunctionCall(id);
            return make_unique<VariableNode>(id);
        }
        if(tok.type == "PAREN" && tok.value=="(") {
            advance();
            auto node = parseExpression();
            expect("PAREN", ")");
            return node;
        }
        if(tok.type == "LCURLY") {
            return parseBlock();
        }
        throw runtime_error("Unexpected token in primary: " + tok.value);
    }
    
    ASTNodePtr parseFunctionCall(const string &funcName) {
        expect("PAREN", "(");
        auto funcCall = make_unique<FunctionCallNode>(funcName);
        if(!(current().type=="PAREN" && current().value==")")) {
            funcCall->addArgument(parseExpression());
            while(current().type=="COMMA") {
                advance();
                funcCall->addArgument(parseExpression());
            }
        }
        expect("PAREN", ")");
        return funcCall;
    }
};

//////////////////////////////////////////////
// MAIN DEMO
//////////////////////////////////////////////

int main() {
    // Example Molang snippet from a Mars worldgen file (extended with assignments, loops, return)
    string molangExample = R"(
        t.biome = (q.noise(v.originx * 0.005, v.originz * 0.005) + 1) * 0.5;
        t.blend = t.biome * t.biome * (3 - 2 * t.biome);
        v.base = 93;
        v.huge = (q.noise(v.originx * 0.003125, v.originz * 0.003125)
                    + 0.25 * q.noise(v.originx * 0.00625, v.originz * 0.00625)
                    + 0.0625 * q.noise(v.originx * 0.0125, v.originz * 0.0125)
                    + 0.015625 * q.noise(v.originx * 0.025, v.originz * 0.025)) * 12;
        loop(4, {
            v.noise1_val = v.noise1_val + q.noise(v.originx * 0.0125, v.originz * 0.0125) * 1;
            v.noise2_val = v.noise2_val + q.noise(-v.originx * 0.015, -v.originz * 0.015) * 1;
        });
        t.height = t.blend * (90 + v.noise1_val) + ((1 - t.blend) * (v.base + v.huge));
        t.layer = (math.floor(t.height) < 92 ? math.floor(t.height) - 92 : 0)
    )";
    
    try {
        auto tokens = tokenize(molangExample);
        Parser parser(tokens);
        auto program = parser.parseProgram();
        json densityJson = program->toDensityJson();
        cout << "=== Converted Density Function JSON ===" << endl;
        cout << densityJson.dump(2) << endl;
    } catch (const exception &e) {
        cerr << "Error: " << e.what() << endl;
        return 1;
    }
    
    // Example of processing a worldgen JSON snippet:
    json worldgenJson = {
      {"format_version", "1.20.20"},
      {"minecraft:scatter_feature", {
          {"description", { {"identifier", "cosmos:mars/base/layer_picker"} }},
          {"places_feature", "cosmos:mars/base/block_picker"},
          {"iterations", "t.height"},
          {"x", 0},
          {"z", "t.layer = t.layer + 1; return 0;"},
          {"y", {
              {"distribution", "fixed_grid"},
              {"extent", json::array({0, "t.height - 1"})}
          }}
      }}
    };
    
    cout << "\n=== Original Worldgen JSON ===" << endl;
    cout << worldgenJson.dump(2) << endl;
    
    
    
    return 0;
}

File: nodeGenerators/Refracted_madness/worldgenconverter.js
"use strict";

/* =========================
   TOKENIZER
========================= */
class Token {
  constructor(type, value) {
    this.type = type;
    this.value = value;
  }
}

function tokenize(s) {
  let tokens = [];
  let i = 0;
  while (i < s.length) {
    let c = s[i];
    if (/\s/.test(c)) {
      i++;
      continue;
    }
    if (/\d/.test(c) || (c === '.' && i + 1 < s.length && /\d/.test(s[i + 1]))) {
      let numStr = c;
      i++;
      while (i < s.length && (/\d/.test(s[i]) || s[i] === '.')) {
        numStr += s[i];
        i++;
      }
      tokens.push(new Token("NUMBER", numStr));
      continue;
    }
    if (/[a-zA-Z_.]/.test(c)) {
      let idStr = c;
      i++;
      while (i < s.length && /[a-zA-Z0-9_.]/.test(s[i])) {
        idStr += s[i];
        i++;
      }
      tokens.push(new Token("IDENTIFIER", idStr));
      continue;
    }
    if ("<>!=".includes(c)) {
      let op = c;
      if (i + 1 < s.length && s[i + 1] === "=") {
        op += "=";
        i += 2;
      } else {
        i++;
      }
      tokens.push(new Token("OPERATOR", op));
      continue;
    }
    if ("+-*/%".includes(c)) {
      tokens.push(new Token("OPERATOR", c));
      i++;
      continue;
    }
    if ("()".includes(c)) {
      tokens.push(new Token("PAREN", c));
      i++;
      continue;
    }
    if (c === ",") {
      tokens.push(new Token("COMMA", ","));
      i++;
      continue;
    }
    if (c === "?") {
      tokens.push(new Token("QUESTION", "?"));
      i++;
      continue;
    }
    if (c === ":") {
      tokens.push(new Token("COLON", ":"));
      i++;
      continue;
    }
    if (c === ";") {
      tokens.push(new Token("SEMICOLON", ";"));
      i++;
      continue;
    }
    if (c === "{") {
      tokens.push(new Token("LCURLY", "{"));
      i++;
      continue;
    }
    if (c === "}") {
      tokens.push(new Token("RCURLY", "}"));
      i++;
      continue;
    }
    // Skip any unrecognized character.
    i++;
  }
  tokens.push(new Token("EOF", ""));
  return tokens;
}

/* =========================
   AST NODE CLASSES
========================= */

class ASTNode {
  toMolang() {
    throw new Error("Not implemented");
  }
  toDensityJson() {
    throw new Error("Not implemented");
  }
}

class NumberNode extends ASTNode {
  constructor(valueStr) {
    super();
    this.value = parseFloat(valueStr);
  }
  toMolang() {
    return this.value.toString();
  }
  toDensityJson() {
    return { type: "minecraft:number", value: this.value };
  }
}

class VariableNode extends ASTNode {
  constructor(name) {
    super();
    this.name = name;
  }
  toMolang() {
    return this.name;
  }
  toDensityJson() {
    return { type: "minecraft:variable", name: this.name };
  }
}

class BinaryOpNode extends ASTNode {
  constructor(op, left, right) {
    super();
    this.op = op;
    this.left = left;
    this.right = right;
  }
  toMolang() {
    return "(" + this.left.toMolang() + " " + this.op + " " + this.right.toMolang() + ")";
  }
  toDensityJson() {
    return {
      type: "minecraft:binary_operation",
      operator: this.op,
      left: this.left.toDensityJson(),
      right: this.right.toDensityJson(),
    };
  }
}

class FunctionCallNode extends ASTNode {
  constructor(functionName) {
    super();
    this.functionName = functionName;
    this.arguments = [];
  }
  addArgument(arg) {
    this.arguments.push(arg);
  }
  toMolang() {
    let argsStr = this.arguments.map(arg => arg.toMolang()).join(", ");
    return this.functionName + "(" + argsStr + ")";
  }
  toDensityJson() {
    // Map specific functions to documented density nodes.
    if (this.functionName === "q.noise" && this.arguments.length === 2) {
      return {
        type: "minecraft:noise",
        x: this.arguments[0].toDensityJson(),
        y: this.arguments[1].toDensityJson(),
      };
    } else if (this.functionName === "math.clamp" && this.arguments.length === 3) {
      return {
        type: "minecraft:clamp",
        input: this.arguments[0].toDensityJson(),
        min: this.arguments[1].toDensityJson(),
        max: this.arguments[2].toDensityJson(),
      };
    } else if (this.functionName === "math.floor" && this.arguments.length === 1) {
      return {
        type: "minecraft:floor",
        input: this.arguments[0].toDensityJson(),
      };
    } else if (this.functionName === "math.max" && this.arguments.length === 2) {
      return {
        type: "minecraft:max",
        a: this.arguments[0].toDensityJson(),
        b: this.arguments[1].toDensityJson(),
      };
    }
    // Otherwise, use a generic function node.
    let argsJson = this.arguments.map(arg => arg.toDensityJson());
    return {
      type: "minecraft:function",
      name: this.functionName,
      args: argsJson,
    };
  }
}

class TernaryOpNode extends ASTNode {
  constructor(condition, trueExpr, falseExpr) {
    super();
    this.condition = condition;
    this.trueExpr = trueExpr;
    this.falseExpr = falseExpr;
  }
  toMolang() {
    return (
      "(" +
      this.condition.toMolang() +
      " ? " +
      this.trueExpr.toMolang() +
      " : " +
      this.falseExpr.toMolang() +
      ")"
    );
  }
  toDensityJson() {
    return {
      type: "minecraft:ternary",
      condition: this.condition.toDensityJson(),
      true: this.trueExpr.toDensityJson(),
      false: this.falseExpr.toDensityJson(),
    };
  }
}

class AssignmentNode extends ASTNode {
  constructor(variable, expression) {
    super();
    this.variable = variable; // VariableNode
    this.expression = expression;
  }
  toMolang() {
    return this.variable.toMolang() + " = " + this.expression.toMolang();
  }
  toDensityJson() {
    return {
      type: "minecraft:assign",
      variable: this.variable.toDensityJson(),
      value: this.expression.toDensityJson(),
    };
  }
}

class ReturnNode extends ASTNode {
  constructor(expression) {
    super();
    this.expression = expression;
  }
  toMolang() {
    return "return " + this.expression.toMolang();
  }
  toDensityJson() {
    return {
      type: "minecraft:return",
      value: this.expression.toDensityJson(),
    };
  }
}

class BlockNode extends ASTNode {
  constructor() {
    super();
    this.statements = [];
  }
  addStatement(stmt) {
    this.statements.push(stmt);
  }
  toMolang() {
    return "{ " + this.statements.map(s => s.toMolang()).join("; ") + " }";
  }
  toDensityJson() {
    return {
      type: "minecraft:block",
      statements: this.statements.map(s => s.toDensityJson()),
    };
  }
}

class LoopNode extends ASTNode {
  constructor(iterations, body) {
    super();
    this.iterations = iterations;
    this.body = body; // BlockNode
  }
  toMolang() {
    return "loop(" + this.iterations.toMolang() + ", " + this.body.toMolang() + ")";
  }
  toDensityJson() {
    return {
      type: "minecraft:loop",
      iterations: this.iterations.toDensityJson(),
      body: this.body.toDensityJson(),
    };
  }
}

/* =========================
   PARSER
========================= */

class Parser {
  constructor(tokens) {
    this.tokens = tokens;
    this.pos = 0;
  }
  current() {
    return this.tokens[this.pos];
  }
  advance() {
    if (this.pos < this.tokens.length - 1) this.pos++;
  }
  expect(type, value = "") {
    let tok = this.current();
    if (tok.type !== type || (value !== "" && tok.value !== value))
      throw new Error(
        "Expected token " +
          type +
          " " +
          value +
          " but got " +
          tok.type +
          " " +
          tok.value
      );
    this.advance();
  }
  // program = { statement ";" }+
  parseProgram() {
    let block = new BlockNode();
    while (this.current().type !== "EOF") {
      let stmt = this.parseStatement();
      block.addStatement(stmt);
      if (this.current().type === "SEMICOLON") this.advance();
    }
    return block;
  }
  // statement: "return" expression | "loop" "(" expression "," block ")" | expression (optionally assignment)
  parseStatement() {
    if (this.current().type === "IDENTIFIER" && this.current().value === "return") {
      this.advance();
      return new ReturnNode(this.parseExpression());
    }
    if (this.current().type === "IDENTIFIER" && this.current().value === "loop") {
      return this.parseLoopStatement();
    }
    let expr = this.parseExpression();
    if (
      expr instanceof VariableNode &&
      this.current().type === "OPERATOR" &&
      this.current().value === "="
    ) {
      let varNode = expr;
      this.advance(); // consume '='
      let rightExpr = this.parseExpression();
      return new AssignmentNode(varNode, rightExpr);
    }
    return expr;
  }
  parseLoopStatement() {
    this.advance(); // consume "loop"
    this.expect("PAREN", "(");
    let iterations = this.parseExpression();
    this.expect("COMMA", ",");
    let body = this.parseBlock();
    this.expect("PAREN", ")");
    return new LoopNode(iterations, body);
  }
  parseBlock() {
    this.expect("LCURLY", "{");
    let block = new BlockNode();
    while (this.current().type !== "RCURLY") {
      let stmt = this.parseStatement();
      block.addStatement(stmt);
      if (this.current().type === "SEMICOLON") this.advance();
    }
    this.expect("RCURLY", "}");
    return block;
  }
  parseExpression() {
    return this.parseTernary();
  }
  parseTernary() {
    let cond = this.parseComparison();
    if (this.current().type === "QUESTION") {
      this.advance();
      let trueExpr = this.parseExpression();
      this.expect("COLON", ":");
      let falseExpr = this.parseExpression();
      return new TernaryOpNode(cond, trueExpr, falseExpr);
    }
    return cond;
  }
  parseComparison() {
    let node = this.parseAdditive();
    while (
      this.current().type === "OPERATOR" &&
      ["<", ">", "<=", ">=", "==", "!="].includes(this.current().value)
    ) {
      let op = this.current().value;
      this.advance();
      let right = this.parseAdditive();
      node = new BinaryOpNode(op, node, right);
    }
    return node;
  }
  parseAdditive() {
    let node = this.parseMultiplicative();
    while (
      this.current().type === "OPERATOR" &&
      (this.current().value === "+" || this.current().value === "-")
    ) {
      let op = this.current().value;
      this.advance();
      let right = this.parseMultiplicative();
      node = new BinaryOpNode(op, node, right);
    }
    return node;
  }
  parseMultiplicative() {
    let node = this.parseUnary();
    while (
      this.current().type === "OPERATOR" &&
      (this.current().value === "*" ||
        this.current().value === "/" ||
        this.current().value === "%")
    ) {
      let op = this.current().value;
      this.advance();
      let right = this.parseUnary();
      node = new BinaryOpNode(op, node, right);
    }
    return node;
  }
  parseUnary() {
    if (
      this.current().type === "OPERATOR" &&
      (this.current().value === "+" || this.current().value === "-")
    ) {
      let op = this.current().value;
      this.advance();
      let node = this.parseUnary();
      return new BinaryOpNode(op, new NumberNode("0"), node);
    }
    return this.parsePrimary();
  }
  parsePrimary() {
    let tok = this.current();
    if (tok.type === "NUMBER") {
      this.advance();
      return new NumberNode(tok.value);
    }
    if (tok.type === "IDENTIFIER") {
      let id = tok.value;
      this.advance();
      if (this.current().type === "PAREN" && this.current().value === "(")
        return this.parseFunctionCall(id);
      return new VariableNode(id);
    }
    if (tok.type === "PAREN" && tok.value === "(") {
      this.advance();
      let node = this.parseExpression();
      this.expect("PAREN", ")");
      return node;
    }
    if (tok.type === "LCURLY") {
      return this.parseBlock();
    }
    throw new Error("Unexpected token in primary: " + tok.value);
  }
  parseFunctionCall(funcName) {
    this.expect("PAREN", "(");
    let funcCall = new FunctionCallNode(funcName);
    if (!(this.current().type === "PAREN" && this.current().value === ")")) {
      funcCall.addArgument(this.parseExpression());
      while (this.current().type === "COMMA") {
        this.advance();
        funcCall.addArgument(this.parseExpression());
      }
    }
    this.expect("PAREN", ")");
    return funcCall;
  }
}

/* =========================
   REVERSE CONVERSION: Density JSON → Molang
   (Based on the node "type" field)
========================= */
function densityJsonToAst(djson) {
  switch (djson.type) {
    case "minecraft:number":
      return new NumberNode(djson.value.toString());
    case "minecraft:variable":
      return new VariableNode(djson.name);
    case "minecraft:binary_operation":
      return new BinaryOpNode(
        djson.operator,
        densityJsonToAst(djson.left),
        densityJsonToAst(djson.right)
      );
    case "minecraft:noise":
      {
        let func = new FunctionCallNode("q.noise");
        func.addArgument(densityJsonToAst(djson.x));
        func.addArgument(densityJsonToAst(djson.y));
        return func;
      }
    case "minecraft:clamp":
      {
        let func = new FunctionCallNode("math.clamp");
        func.addArgument(densityJsonToAst(djson.input));
        func.addArgument(densityJsonToAst(djson.min));
        func.addArgument(densityJsonToAst(djson.max));
        return func;
      }
    case "minecraft:floor":
      {
        let func = new FunctionCallNode("math.floor");
        func.addArgument(densityJsonToAst(djson.input));
        return func;
      }
    case "minecraft:max":
      {
        let func = new FunctionCallNode("math.max");
        func.addArgument(densityJsonToAst(djson.a));
        func.addArgument(densityJsonToAst(djson.b));
        return func;
      }
    case "minecraft:ternary":
      return new TernaryOpNode(
        densityJsonToAst(djson.condition),
        densityJsonToAst(djson.true),
        densityJsonToAst(djson.false)
      );
    case "minecraft:assign":
      return new AssignmentNode(
        densityJsonToAst(djson.variable),
        densityJsonToAst(djson.value)
      );
    case "minecraft:return":
      return new ReturnNode(densityJsonToAst(djson.value));
    case "minecraft:block":
      {
        let block = new BlockNode();
        for (let stmt of djson.statements) {
          block.addStatement(densityJsonToAst(stmt));
        }
        return block;
      }
    case "minecraft:loop":
      return new LoopNode(
        densityJsonToAst(djson.iterations),
        densityJsonToAst(djson.body)
      );
    case "minecraft:function":
      {
        let func = new FunctionCallNode(djson.name);
        for (let arg of djson.args) {
          func.addArgument(densityJsonToAst(arg));
        }
        return func;
      }
    default:
      throw new Error("Unknown density function node type: " + djson.type);
  }
}

function densityToMolang(djson) {
  let ast = densityJsonToAst(djson);
  return ast.toMolang();
}

/* =========================
   EXPORTS
========================= */
module.exports = {
  molangToDensity: function (molangStr) {
    let tokens = tokenize(molangStr);
    let parser = new Parser(tokens);
    let program = parser.parseProgram();
    return program.toDensityJson();
  },
  densityToMolang: densityToMolang,
};
